[InstCombine] handle (X + A) % Op1 for small X, A

DesWurstes opened the pull request:
Allows LLVM to optimize rem out of loops like this:

```c
void c(void)
{
 unsigned m = 0;
 for(int i = 0; i < x; i++)
 {
 m += 7;
 m %= 600;
 g(m);
 }
}
```

by replacing it with `m + 7 >= 600 ? m + 7 - 600 : m + 7`

Alive2 proof: https://alive2.llvm.org/ce/z/NXHJJD


```llvm
define noundef i16 @src(i16 noundef %x, i16 noundef %n, i16 noundef %a) unnamed_addr  {
start:
  %_3 = icmp ult i16 %x, %n
  tail call void @llvm.assume(i1 %_3)
  %_4 = icmp ult i16 %a, %n
  tail call void @llvm.assume(i1 %_4)
  %_6 = add nuw i16 %x, %a
  %0 = urem i16 %_6, %n
  ret i16 %0
}

define noundef i16 @tgt(i16 noundef %x, i16 noundef %n, i16 noundef %a) unnamed_addr  {
start:
  %_3 = icmp ult i16 %x, %n
  tail call void @llvm.assume(i1 %_3)
  %_4 = icmp ult i16 %a, %n
  tail call void @llvm.assume(i1 %_4)
  %_6 = add nuw i16 %x, %a
  %w = icmp uge i16 %_6, %n
  %_7 = sub nuw i16 %_6, %n
  %0 = select i1 %w, i16 %_7, i16 %_6
  ret i16 %0
}

declare void @llvm.assume(i1 noundef) 
```

I added the equivalents of tests for `(X + 1) % Op1`, except for `urem_with_dominating_condition` and `urem_with_dominating_condition_false` which don't get matched by the pattern matcher and don't get optimized. I couldn't find a reason for this in my code, the reason is likely outside it. If I were to add these, these would be:

```llvm
; https://alive2.llvm.org/ce/z/Y6jvTf
define i8 @urem_with_dominating_condition_a(i8 %x, i8 %n, i8 %a) {
start:
  %cond = icmp ult i8 %x, %n
  br i1 %cond, label %.bb0, label %.bb2
.bb0:
  %cond2 = icmp ult i8 %a, %n
  br i1 %cond2, label %.bb1, label %.bb2
.bb1:
  %add = add i8 %x, %a
  %out = urem i8 %add, %n
  ret i8 %out
.bb2:
  ret i8 0
}

; Revert the dominating condition and target branch at the same time.
define i8 @urem_with_dominating_condition_false_a(i8 %x, i8 %n, i8 %a) {
start:
  %cond = icmp uge i8 %x, %n
  br i1 %cond, label %.bb2, label %.bb0 ; Swap the branch targets
.bb0:
  %cond2 = icmp uge i8 %a, %n
  br i1 %cond2, label %.bb2, label %.bb1 ; Swap the branch targets
.bb1:
  %add = add i8 %x, %a
  %out = urem i8 %add, %n
  ret i8 %out
.bb2:
  ret i8 0
}
```

DesWurstes commented:
I'll fix my commit email address and reopen

DesWurstes committed:
[test] precommit some tests for (X+A)%Op1

DesWurstes committed:
[InstCombine] handle (X + A) % Op1 for small X, A

DesWurstes commented:
`make check-llvm` does not show any regressions for me, do you know why it might be failing @nikic? Thanks for your time

dtcxzyw commented:
> `make check-llvm` does not show any regressions for me, do you know why it might be failing @nikic? Thanks for your time

The CI failure looks unrelated (flang build failure).

BTW, I think this patch should not be implemented in InstCombine. Generally, InstCombine canonicalizes patterns into a smaller one for better analysis results. BTW, this fold is not always profitable due to the following two reasons:
1. Div/Rem are faster than you expected. On some microarchitectures, the latency scales linearly with `Log2(Dividend) - Log2(Divisor)`. If the divisor is known at compile-time, the compiler may expand div/rem into mul+add sequences.
2. For some backends without CMOV instructions, the branch misprediction penalty may have a large impact on performance.

Therefore, I think you can implement this transformation in some target-dependent passes (e.g., CGP).
Additionally, for your motivating case, the branch misprediction penalty is negligible because `m` is an induction variable of a loop. We have implemented this optimization in https://github.com/llvm/llvm-project/pull/104724. You can generalize the PR if it is not enough to satisfy your demand :)

DesWurstes commented:
Okay, I had implemented it there because just above where I added my code, there's optimization for `// For "(X + 1) % Op1" and if (X u< Op1) => (X + 1) == Op1 ? 0 : X + 1 .`. I'll try to improve the code in the pull request you've linked
